internal/utils/utils.go

```diff
@@ -0,0 +1,363 @@
+	sts := &appsv1.StatefulSet{}
+	err := c.Get(ctx, client.ObjectKey{Namespace: ec.Namespace, Name: ec.Name}, sts)
+	if err != nil {
+		return err
+	}
+	if !metav1.IsControlledBy(sts, ec) {
+		return fmt.Errorf("StatefulSet %s/%s is not controlled by EtcdCluster %s/%s", sts.Namespace, sts.Name, ec.Namespace, ec.Name)
+	}
+	return nil
+}
+
+func configMapNameForEtcdCluster(ec *ecv1alpha1.EtcdCluster) string {
+	return fmt.Sprintf("%s-state", ec.Name)
+}
+
+func PeerEndpointForOrdinalIndex(ec *ecv1alpha1.EtcdCluster, index int) (string, string) {
+	name := fmt.Sprintf("%s-%d", ec.Name, index)
+	return name, fmt.Sprintf("http://%s-%d.%s.%s.svc.cluster.local:2380",
+		ec.Name, index, ec.Name, ec.Namespace)
+}
+
+func newEtcdClusterState(ec *ecv1alpha1.EtcdCluster, replica int) *corev1.ConfigMap {
+	// We always add members one by one, so the state is always
+	// "existing" if replica > 1.
+	state := "new"
+	if replica > 1 {
+		state = "existing"
+	}
+
+	var initialCluster []string
+	for i := 0; i < replica; i++ {
+		name, peerURL := PeerEndpointForOrdinalIndex(ec, i)
+		initialCluster = append(initialCluster, fmt.Sprintf("%s=%s", name, peerURL))
+	}
+
+	return &corev1.ConfigMap{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configMapNameForEtcdCluster(ec),
+			Namespace: ec.Namespace,
+		},
+		Data: map[string]string{
+			"ETCD_INITIAL_CLUSTER_STATE": state,
+			"ETCD_INITIAL_CLUSTER":       strings.Join(initialCluster, ","),
+		},
+	}
+}
+
+func applyEtcdClusterState(ctx context.Context, ec *ecv1alpha1.EtcdCluster, replica int, c client.Client, scheme *runtime.Scheme) error {
+	cm := newEtcdClusterState(ec, replica)
+
+	err := c.Get(ctx, types.NamespacedName{Name: configMapNameForEtcdCluster(ec), Namespace: ec.Namespace}, &corev1.ConfigMap{})
+	if err != nil && errors.IsNotFound(err) {
+		createErr := c.Create(ctx, cm)
+		return createErr
+	}
+
+	if err != nil && !errors.IsNotFound(err) {
+		return fmt.Errorf("cannot find ConfigMap for EtcdCluster %s: %w", ec.Name, err)
+	}
+
+	if err := validateOwner(ec, cm); err != nil {
+		return err
+	}
+	// Create a new controller ref.
+	owners, err := prepareOwnerReference(ec, scheme)
+	if err != nil {
+		return err
+	}
+
+	cm.OwnerReferences = owners
+
+	updateErr := c.Update(ctx, cm)
+	return updateErr
+}
+
+func IsStatefulSetDesiredSize(ctx context.Context, etcdCluster *ecv1alpha1.EtcdCluster, r client.Client) (string, error) {
+	sts := &appsv1.StatefulSet{}
+	err := r.Get(ctx, client.ObjectKey{Name: etcdCluster.Name, Namespace: etcdCluster.Namespace}, sts)
+	if err != nil {
+		return "", err
+	}
+	if *sts.Spec.Replicas < int32(etcdCluster.Spec.Size) {
+		return SSSmallerThanDesired, nil
+	}
+
+	if *sts.Spec.Replicas > int32(etcdCluster.Spec.Size) {
+		return SSBiggerThanDesired, nil
+	}
+
+	if *sts.Spec.Replicas == int32(etcdCluster.Spec.Size) {
+		return SSEqualToDesired, nil
+	}
+
+	return "", nil
+}
```

Comment:
It's a little over engineering to me. I think we just need to provide a function to get the statefulset.